xclim/sdba/adjustment.py

```diff
@@ -406,6 +459,168 @@ def _adjust(self, sim, interp="nearest", extrapolation="constant"):
+
+        Parameters
+        ----------
+        At instantiation:
+
+        cluster_thresh: Quantity (str with units)
+          The threshold value for defining clusters.
+        q_thresh : float
+          The quantile of "extreme" values, [0, 1[.
+        dist : str
+          Name of the distribution (scipy name) to use.
+
+        In adjustment:
+
+        frac: float
+          Fraction where the cutoff happens between the original scen and the corrected one.
+          See Notes, ]0, 1].
+        power: float
+          Shape of the correction strength, see Notes.
+
+        Notes
+        -----
+        Extreme values are extracted from `ref`, `hist` and `sim` by finding all "clusters",
+        i.e. runs of consecutive values above `cluster_thresh`. The `q_thresh`th percentile
+        of these values is taken on `ref` and `hist` and becomes `thresh`, the extreme value
+        threshold. The maximal value of each cluster of `ref`, if it exceeds that new threshold,
+        is taken and a `dist` (default : Generalized Pareto) distribution is fitted to them.
+        Similarly with `sim`. The cdf of the extreme values of `sim` is computed in reference
+        to the distribution fitted on `sim` and then the corresponding values (quantile / ppf)
+        in reference to the distribution fitted on `ref` are taken as the new bias-adjusted values.
+
+        Once new extreme values are found, a mixture from the original scen and corrected scen
+        is used in the result. For each original value :math:`S_i` and corrected value :math:`C_i`
+        the final extreme value :math:`V_i` is:
+
+        .. math::
+
+            V_i = C_i * \tau + S_i * (1 - \tau)
+
+        Where :math:`\tau` is a function of sim's extreme values :math:`F` and of arguments
+        ``frac`` (:math:`f`) and ``power`` (:math:`p`):
+
+        .. math::
+
+            \tau = \left(\frac{1}{f}\frac{S - min(S)}{max(S) - min(S)}\right)^p
+
+        References
+        ----------
+        .. [ClimateTools] https://juliaclimate.github.io/ClimateTools.jl/stable/
+        .. [RRJF2021] Roy, P., Rondeau-Genesse, G., Jalbert, J., Fournier, É. 2021. Climate Scenarios of Extreme Precipitation Using a Combination of Parametric and Non-Parametric Bias Correction Methods. Submitted to Climate Services, April 2021.
+        """
+        super().__init__(q_thresh=q_thresh, cluster_thresh=cluster_thresh, dist=dist)
+
+    def _train(self, ref, hist):
+        cluster_thresh = convert_units_to(self.cluster_thresh, ref)
+        hist = convert_units_to(hist, ref)
+
+        # Extreme value threshold computed relative to "large values".
+        # We use the mean between ref and hist here.
+        thresh = (
+            ref.where(ref >= cluster_thresh).quantile(self.q_thresh, dim="time")
+            + hist.where(hist >= cluster_thresh).quantile(self.q_thresh, dim="time")
+        ) / 2
```

Comment:
Hum, our definition of "hist" is the timeseries of _raw_ model data, so there might be a difference between the two thresh.

There's some ambiguity here on whether `ExtremeValue.train` happens after or before the "first-order" Adjustment.

Reply:
Right, if already adjusted. The original algo estimate it on non-adjusted vector. Either way should do the trick, since minor difference will not affect significantly the convergence (I think!).

Reply:
But, is there a reason that the thresh is not recomputed for the future data? I mean, this way the "extreme values" of a given array would be those with  `quantile(arr[arr > cluster_thresh]) > q_thresh`.

Here we transform the quantile thresh into a data thresh on ref+hist, and then use it on sim. Could make sense to stick to the quantile thresh on both?

Reply:
What you modelize with the GPD is an exceedence distribution over a given threshold, not over a quantile. So that's why the threshold remain constant. You could however modelize the evolution of GPD parameters, but this is somewhat more hard to do and clearly beyond the scope of the developed method. 

However, it is easier to develop a non-stationairy approach with GEV parameters and this could be used as input more easily and is the next phases I was envisioning when we did the method.